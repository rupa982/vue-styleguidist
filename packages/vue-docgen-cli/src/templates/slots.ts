import { SlotDescriptor } from 'vue-docgen-api'
import { cleanReturn } from './utils'

export default (slots: { [slotName: string]: SlotDescriptor }): string => {
	const slotNames = Object.keys(slots).filter(slotName => slots[slotName].description)
	if (!slotNames.length) {
		return ''
	}
	return `
## Slots

| Name          | Description  | Bindings |
| ------------- | ------------ | -------- |
${slotNames
	.map(slotName => {
		const { description, bindings } = slots[slotName]
		const readableBindings = // serialize bindings to display them ina readable manner
			bindings && Object.keys(bindings).length ? JSON.stringify(bindings, null, 2) : ''
		return cleanReturn(`| ${slotName} | ${description} | ${readableBindings} |`) // remplace returns by <br> to allow them in a table cell
	})
	.join('\n')}
`
}
