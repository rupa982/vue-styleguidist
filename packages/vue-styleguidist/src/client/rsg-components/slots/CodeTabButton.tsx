import React from 'react'
import PropTypes from 'prop-types'
import TabButton from 'rsg-components/TabButton'

interface CodeTabeButtonProps {
	onClick: (e: React.MouseEvent) => void
	name: string
	active?: boolean
}

const CodeTabButton = (props: CodeTabeButtonProps) => <TabButton {...props}>View Code</TabButton>

CodeTabButton.propTypes = {
	onClick: PropTypes.func.isRequired,
	name: PropTypes.string.isRequired,
	active: PropTypes.bool
}

export default CodeTabButton
