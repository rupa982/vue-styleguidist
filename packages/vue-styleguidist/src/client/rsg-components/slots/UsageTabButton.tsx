import React from 'react'
import PropTypes from 'prop-types'
import TabButton from 'rsg-components/TabButton'
import isEmpty from 'lodash/isEmpty'
import { PropDescriptor, MethodDescriptor, EventDescriptor, SlotDescriptor } from 'vue-docgen-api'

export interface UsageTabButtonProps {
	name: string
	onClick: (e: React.MouseEvent) => void
	active?: boolean
	props: {
		props?: PropDescriptor[]
		methods?: MethodDescriptor[]
		events?: EventDescriptor[]
		slots?: SlotDescriptor[]
	}
}

const UsageTabButton = (props: UsageTabButtonProps) => {
	const component = props.props
	const showButton =
		!isEmpty(component.props) ||
		!isEmpty(component.methods) ||
		!isEmpty(component.events) ||
		!isEmpty(component.slots)
	return showButton ? <TabButton {...props}>Props, methods, events & slots</TabButton> : null
}

UsageTabButton.propTypes = {
	onClick: PropTypes.func.isRequired,
	name: PropTypes.string.isRequired,
	props: PropTypes.shape({
		props: PropTypes.array,
		methods: PropTypes.array
	}).isRequired,
	active: PropTypes.bool
}

export default UsageTabButton
