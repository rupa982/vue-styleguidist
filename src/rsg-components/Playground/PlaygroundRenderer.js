import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import Styled from 'rsg-components/Styled';

export const styles = ({ space, color, borderRadius }) => ({
	root: {
		marginBottom: space[4],
	},
	preview: {
		padding: space[2],
		border: [[1, color.border, 'solid']],
		borderRadius,
	},
	controls: {
		display: 'flex',
		alignItems: 'center',
	},
	toolbar: {
		marginLeft: 'auto',
	},
	tab: {}, // expose className to allow using it in 'styles' settings
});

export function PlaygroundRenderer(
	{ classes, name, preview, previewProps, tabButtons, tabBody, toolbar },
	{ config }
) {
	const { className, ...props } = previewProps;
	const navigation = config.navigation;
	return (
		<div className={classes.root}>
			<div className={cx(classes.preview, className)} {...props} data-preview={name}>
				{preview}
			</div>
			<div className={classes.controls}>
				<div className={classes.tabs}>{tabButtons}</div>
				{!navigation ? <div className={classes.toolbar}>{toolbar}</div> : null}
			</div>
			<div className={classes.tab}>{tabBody}</div>
		</div>
	);
}

PlaygroundRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	name: PropTypes.string.isRequired,
	preview: PropTypes.node.isRequired,
	previewProps: PropTypes.object.isRequired,
	tabButtons: PropTypes.node.isRequired,
	tabBody: PropTypes.node.isRequired,
	toolbar: PropTypes.node.isRequired,
};

PlaygroundRenderer.contextTypes = {
	config: PropTypes.object,
};

export default Styled(styles)(PlaygroundRenderer);
