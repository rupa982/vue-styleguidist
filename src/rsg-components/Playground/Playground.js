import React, { Component } from 'react';
import PropTypes from 'prop-types';
import debounce from 'lodash/debounce';
import Preview from 'rsg-components/Preview';
import Para from 'rsg-components/Para';
import Slot from 'rsg-components/Slot';
import PlaygroundRenderer from 'rsg-components/Playground/PlaygroundRenderer';
import { EXAMPLE_TAB_CODE_EDITOR } from '../slots';

export default class Playground extends Component {
	static propTypes = {
		code: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
		vuex: PropTypes.object,
		index: PropTypes.number.isRequired,
		name: PropTypes.string.isRequired,
		settings: PropTypes.object,
	};

	static contextTypes = {
		config: PropTypes.object.isRequired,
		isolatedExample: PropTypes.bool,
	};

	constructor(props, context) {
		super(props, context);
		const { code } = props;
		const { config } = context;

		this.showCode = config.showCode;
		this.handleChange = this.handleChange.bind(this);
		this.handleTabChange = this.handleTabChange.bind(this);
		this.handleChange = debounce(this.handleChange, config.previewDelay);

		this.state = {
			code,
			activeTab: undefined,
		};
	}

	componentDidMount() {
		if (this.showCode) {
			setTimeout(() => {
				this.setState({
					activeTab: EXAMPLE_TAB_CODE_EDITOR,
				});
			}, 0);
		}
	}

	componentWillReceiveProps(nextProps) {
		const { code } = nextProps;
		this.setState({
			code,
		});
	}

	shouldComponentUpdate(nextProps, nextState) {
		return nextState.code !== this.state.code || nextState.activeTab !== this.state.activeTab;
	}

	componentWillUnmount() {
		// Clear pending changes
		this.handleChange.cancel();
	}

	handleChange(code) {
		this.setState({
			code,
		});
	}

	handleTabChange(name) {
		this.setState(state => ({
			activeTab: state.activeTab !== name ? name : undefined,
		}));
	}

	render() {
		const { code, activeTab } = this.state;
		const { evalInContext, index, name, vuex, settings } = this.props;
		const { isolatedExample } = this.context;
		const preview = <Preview code={code} vuex={vuex} evalInContext={evalInContext} />;
		if (settings.noeditor) {
			return (
				<Para>
					{preview}
				</Para>
			);
		}
		if (settings.classname) {
			settings.props = {
				className: settings.classname,
			};
		}
		return (
			<PlaygroundRenderer
				name={name}
				preview={preview}
				previewProps={settings.props || {}}
				tabButtons={
					<Slot
						name="exampleTabButtons"
						active={activeTab}
						props={{ onClick: this.handleTabChange }}
					/>
				}
				tabBody={
					<Slot
						name="exampleTabs"
						active={activeTab}
						onlyActive
						props={{ code, onChange: this.handleChange }}
					/>
				}
				toolbar={
					<Slot name="exampleToolbar" props={{ name, isolated: isolatedExample, example: index }} />
				}
			/>
		);
	}
}
