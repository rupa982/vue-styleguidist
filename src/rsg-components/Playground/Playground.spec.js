import React from 'react';
import PropTypes from 'prop-types';
import Playground from './Playground';
import '../../styles/setupjss';
import slots, { EXAMPLE_TAB_CODE_EDITOR } from '../slots';
import { PlaygroundRenderer } from './PlaygroundRenderer';

const evalInContext = a =>
	new Function('require', 'const React = require("react");' + a).bind(null, require); // eslint-disable-line no-new-func
const code = '<button>OK</button>';
const newCode = '<button>Not OK</button>';
const options = {
	context: {
		config: {
			showCode: false,
			highlightTheme: 'base16-light',
		},
		slots,
	},
	childContextTypes: {
		slots: PropTypes.object.isRequired,
	},
};

it('should render component renderer', () => {
	const actual = shallow(
		<Playground code={code} evalInContext={evalInContext} name="name" index={0} />,
		options
	);

	expect(actual).toMatchSnapshot();
});

it('should update code via props', () => {
	const actual = shallow(
		<Playground code={code} evalInContext={evalInContext} name="name" index={0} />,
		options
	);

	expect(actual.state('code')).toEqual(code);

	actual.setProps({
		code: newCode,
	});

	expect(actual.state('code')).toEqual(newCode);
});

it('should update code with debounce', done => {
	const actual = shallow(
		<Playground code={code} evalInContext={evalInContext} name="name" index={0} />,
		{
			context: {
				...options.context,
				config: {
					...options.context.config,
					previewDelay: 1,
				},
			},
		}
	);

	expect(actual.state('code')).toEqual(code);

	actual.instance().handleChange(newCode);

	expect(actual.state('code')).toEqual(code);
	setTimeout(() => {
		expect(actual.state('code')).toEqual(newCode);
		done();
	}, 3);
});

it('should open a code editor', () => {
	const actual = mount(
		<Playground code={code} evalInContext={evalInContext} name="name" index={0} />,
		options
	);

	expect(actual.find('.ReactCodeMirror')).toHaveLength(0);

	actual.find(`button[name="${EXAMPLE_TAB_CODE_EDITOR}"]`).simulate('click');

	expect(actual.find('.ReactCodeMirror')).toHaveLength(1);
});

it('renderer should render preview', () => {
	const actual = shallow(
		<PlaygroundRenderer
			classes={{}}
			name="name"
			preview={<div>preview</div>}
			tabButtons={<div>tab buttons</div>}
			tabBody={<div>tab body</div>}
			toolbar={<div>toolbar</div>}
		/>
	);

	expect(actual).toMatchSnapshot();
});
