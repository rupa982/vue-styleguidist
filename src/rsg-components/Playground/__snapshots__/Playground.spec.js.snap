// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render preview 1`] = `
<div>
  <div
    data-preview="name"
  >
    <div>
      preview
    </div>
  </div>
  <div>
    <div>
      <div>
        tab buttons
      </div>
    </div>
    <div>
      <div>
        toolbar
      </div>
    </div>
  </div>
  <div>
    <div>
      tab body
    </div>
  </div>
</div>
`;

exports[`should render component renderer 1`] = `
<Styled(Playground)
  name="name"
  preview={
    <Preview
      code="<button>OK</button>"
      evalInContext={[Function]}
    />
  }
  tabBody={
    <Slot
      name="exampleTabs"
      onlyActive={true}
      props={
        Object {
          "code": "<button>OK</button>",
          "onChange": [Function],
        }
      }
    />
  }
  tabButtons={
    <Slot
      name="exampleTabButtons"
      props={
        Object {
          "onClick": [Function],
        }
      }
    />
  }
  toolbar={
    <Slot
      name="exampleToolbar"
      props={
        Object {
          "example": 0,
          "isolated": undefined,
          "name": "name",
        }
      }
    />
  }
/>
`;
